onboarding_steps: Add 'onboarding_video' for new users.

prakhar1144 opened the pull request:
PR to add onboarding video for new users.


Fixes: #29304 

<img width="512" alt="Screenshot 2025-03-03 at 3 28 24 PM" src="https://github.com/user-attachments/assets/90771e8f-eab1-4a1f-83df-cac0703fa83b" />

<img width="512" alt="Screenshot 2025-03-03 at 3 28 43 PM" src="https://github.com/user-attachments/assets/0f14b5b8-01e8-45a0-bb48-d179dc1c24f7" />



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on web/src/dialog_widget.ts:
```diff
@@ -174,6 +175,7 @@ export function launch(conf: DialogWidgetConfig): string {
     // * always_visible_scrollbar: Whether the scrollbar is always visible if modal body
     //   has scrollable content. Default behaviour is to hide the scrollbar when it is
     //   not in use.
+    // * ignore_overlay_click: Whether clicking on the overlay will not close the modal.
```
I think we want a name that's more explicit about it closing the modal. Perhaps ideal would be `close_on_overlay_click`, and having it default to true? I'm also not sure if it's accurate to call it `close_on_background_click`, but that would be even more intuitive.

timabbott commented on web/e2e-tests/realm-creation.test.ts:
```diff
@@ -78,6 +78,11 @@ async function realm_creation_tests(page: Page): Promise<void> {
+    // Close it to avoid blocking further test execution.
+    await page.waitForSelector("#onboarding-video-modal", {visible: true});
+    await page.click("#onboarding-video-modal .modal__close");
+
```
We might want to have `populate_db` mark all the `OnboardingSteps` for the initial onboarding things for most of the original users; maybe excluding a couple accounts (say, Aaron and Zoe); I feel like the "Welcome to inbox" type modals usually not helpful when doing development environment work, and that may apply here too.

prakhar1144 replied:
> We might want to have populate_db mark all the OnboardingSteps for the initial onboarding things for most of the original users;

Done

> maybe excluding a couple accounts (say, Aaron and Zoe)

I think the better approach is mark it as seen for everyone. When you want to test these onboarding steps, create a new user. Logically, it makes sense as onboarding steps are for new users.

There's a very low chance that someone will see the `populate_db` code to figure out that these two users have onboarding steps NOT marked as seen.

timabbott replied:
Yeah, that sounds good, the "new realm" button works if one wants to see them again.

timabbott commented on web/src/onboarding_steps.ts:
```diff
@@ -57,10 +61,31 @@ function narrow_to_dm_with_welcome_bot_new_user(
+function show_onboarding_video(): void {
+    if (ONE_TIME_NOTICES_TO_DISPLAY.has("onboarding_video")) {
+        const html_body = render_onboarding_video_modal({
+            video_src: "../../static/videos/onboarding-video.mp4",
```
I'm not sure whether this path is available in a production release tarball; we should check that (probably `tar -tf` on something from download.zulip.com is a fine way to check?).

Hmm, I'm also not sure the `../../` pathing is safe; I think it would be better to write just `/static/...` for a relative path? I guess we have `../../static/` paths in existing templates for accessing emoji, so maybe it's fine.

prakhar1144 replied:
> I'm not sure whether this path is available in a production release tarball; we should check that (probably tar -tf on something from download.zulip.com is a fine way to check?).

Verified. It's present.

timabbott replied:
This should be getting the video from `ONBOARDING_VIDEO_URL`, right?

prakhar1144 replied:
Yeah, right.

Added:
```py
        if settings.ONBOARDING_VIDEO_URL is not None:
            state["onboarding_video_url"] = settings.ONBOARDING_VIDEO_URL
```

timabbott commented on zilencer/management/commands/populate_db.py:
```diff
@@ -946,8 +946,12 @@ def assign_time_zone_by_delivery_email(delivery_email: str, new_time_zone: str)
 
         # Existing users shouldn't narrow to DM with welcome bot on first login.
+        # Also, onboarding video shouldn't be presented to them.
         OnboardingStep.objects.bulk_create(onboarding_steps)
```
Probably we should redo these comments to not list all of the items -- just say something like:

"Mark onboarding steps as seen for existing users to avoid unnecessary popups during development."

prakhar1144 replied:
Done.

alya commented:
When the modal opens, a small version of it flashes by first. We should make sure this doesn't happen.

prakhar1144 commented on web/src/modals.ts:
```diff
@@ -152,6 +152,10 @@ export function open(
         }
 
+        if ($(e.target).is("[data-ignore-overlay-click]")) {
+            return;
```
Kept it in this format instead of

```
        if (!$(e.target).is("[data-close-on-overlay-click]")) {
            return;
         }
```

in this case, we'll have to add `data-close-on-overlay-click` at all those `.hbs` files where micomodal is used.

timabbott replied:
``` diff
diff --git a/web/src/modals.ts b/web/src/modals.ts
index 813a22fae8..c35ee88d8f 100644
--- a/web/src/modals.ts
+++ b/web/src/modals.ts
@@ -153,7 +153,7 @@ export function open(
             return;
         }
 
-        if ($(e.target).is("[ignore-overlay-click]")) {
+        if ($(e.target).hasClass("ignore-overlay-click")) {
             return;
         }
 
diff --git a/web/templates/dialog_widget.hbs b/web/templates/dialog_widget.hbs
index fc230ec21e..c4f8c178ae 100644
--- a/web/templates/dialog_widget.hbs
+++ b/web/templates/dialog_widget.hbs
@@ -1,5 +1,5 @@
 <div class="micromodal" id="{{modal_unique_id}}" aria-hidden="true">
-    <div class="modal__overlay" tabindex="-1" {{#unless close_on_overlay_click}}ignore-overlay-click{{/unless}}>
+    <div class="modal__overlay {{#unless close_on_overlay_click}}ignore-overlay-click{{/unless}}" tabindex="-1">
         <div {{#if id}}id="{{id}}" {{/if}}class="modal__container" role="dialog" aria-modal="true" aria-labelledby="dialog_title">
             <header class="modal__header">
                 <h1 class="modal__title dialog_heading">
```

prakhar1144 commented on web/styles/modal.css:
```diff
@@ -554,3 +554,17 @@
+
+    video {
+        max-width: 100%;
+        cursor: pointer;
```
This is an addition since yesterday, because Alya mentioned we don't plan to autoplay.

```diff
+cursor: pointer
```

prakhar1144 commented:
Updates in the recent push:
* Fixed a bug in the first commit.
* Addressed review comments.

Continuing the discussion related to [Alya's feedback](https://github.com/zulip/zulip/pull/32074#issuecomment-2431122806) in [the #design thread on CZO.](https://chat.zulip.org/#narrow/channel/101-design/topic/onboarding.20video/near/1967721)

prakhar1144 commented:
@alya Can you take a look

> Hm, given that we can't play sound, perhaps we should just have a "play" button for users to click when ready, rather than auto-playing. We'd want a nice, bug button in the middle of the video.

<details><summary>with play button</summary>
<img width="1512" alt="Screenshot 2024-10-24 at 2 01 23 PM" src="https://github.com/user-attachments/assets/2794969b-8361-418a-a5e0-df8a12d8537d">

</details> 

> When the modal opens, a small version of it flashes by first. We should make sure this doesn't happen.

Can't replicate

<details><summary>video</summary>

https://github.com/user-attachments/assets/f2830288-c1c9-479b-805e-798c20b3a488


</details>

alya commented:
The play button screenshot looks good.

alya commented:
You screen capture is different from what I was doing... I just clicked the "Create new user" button in `dev`; I didn't hide/open the browser window.

prakhar1144 commented:
The latest push is fixing responsiveness + adds grey border as discussed in the CZO thread.

timabbott commented:
Let's get this rebased.

I think we need a /etc/zulip/settings.py server setting for what the URL is. `ONBOARDING_VIDEO_URL` will do as the name. `None` should disable this feature; otherwise it's the URL of the video to use.

prakhar1144 commented:
Updates in the recent push:
* rebased to fix conflicts
* Added `ONBOARDING_VIDEO_URL` as [suggested](https://github.com/zulip/zulip/pull/32074#issuecomment-2683612544).
* https://github.com/zulip/zulip/pull/32074#issuecomment-2431124350
   > When the modal opens, a small version of it flashes by first. We should make sure this doesn't happen.
   
   I fixed it. But I'd highlight one thing. **For a fraction of a second**, we'll see the video player with video still loading from it's source. I took a screenshot of that particular frame:
   
<img width="997" alt="Screenshot 2025-03-03 at 3 16 45 PM" src="https://github.com/user-attachments/assets/35c9f513-09d2-4c7d-8582-bdeb660cc1ef" />

@alya  Let me know if the UX seems off due to this...

prakhar1144 commented on zproject/default_settings.py:
```diff
@@ -708,3 +708,7 @@
+
+# URL of the onboarding video displayed to new users.
+# Set it to None to disable it.
+ONBOARDING_VIDEO_URL: str | None = "/static/videos/onboarding-video.mp4"
```
We'll update this URL as per our final decision in [#design > onboarding video: distribution mechanism](https://chat.zulip.org/#narrow/channel/101-design/topic/onboarding.20video.3A.20distribution.20mechanism/with/2105059)

prakhar1144 replied:
Done.

timabbott commented on web/templates/dialog_widget.hbs:
```diff
@@ -1,5 +1,5 @@
 <div class="micromodal" id="{{modal_unique_id}}" aria-hidden="true">
-    <div class="modal__overlay" tabindex="-1">
+    <div class="modal__overlay" tabindex="-1" {{#unless close_on_overlay_click}}data-ignore-overlay-click{{/unless}}>
```
This is a weird name; usually `data-` properties should have a value. I think just calling this `ignore-overlay-click` would be correct.

prakhar1144 replied:
Made it `ignore-overlay-click`

btw, we have things like `data-micromodal-close`, `data-simplebar`. Maybe we can update those post 10.0

alya commented:
I dunno if it's critical, but why does clicking on a future time in the time bar not work for jumping to that time?

alya commented:
OK, here's a next UX iteration for us to try:

## Button UI
- Change the "Skip video" button to grey (like Watch later). We aren't trying to encourage the user to click it.
- Once you start the video (or maybe once you're 30 sec in, if we can do that?), relabel the "Skip video -- ... " button to just "Skip the rest".

## End of video experience
- We need a big CTA in the center of the video area to close the modal and move on to using Zulip once you've finished watching the video. Just "Let's go!" is fine for now; we can iterate.
- Ideally, there'd be a smaller "Replay video" button below the main CTA.
- It might be good to hide the two buttons at the bottom of the modal at this stage, as they don't really make sense anymore.

## Welcome Bot
- Add to the end of its message (new paragraph):

> You can always come back to the [Welcome to Zulip video]() for a quick app overview

![Screenshot 2025-03-03 at 11 34 40@2x](https://github.com/user-attachments/assets/2d4d0e2b-bc4d-4203-b0e7-d2798dbdf3db)

I think it works fine in both admin and non-admin versions.

## Button behavior

- Let's try having the "Watch later" button mark the onboarding step as done (so it won't pop up the video again). Instead, it can trigger a Welcome Bot message in 24 hrs (exact timing TBD):

> Is this a good time to watch the [Welcome to Zulip video]()?
- The `x` button should probably behave the same as the "Skip video" button.

alya commented:
I chatted with @timabbott when drafting the proposal above. We can try it and iterate more if needed here, and I'll post in  #design for general feedback once I'm feeling pretty happy with where we are.

alya commented:
You can also post there if some of this is tricky to implement and we should think about adjusting the plan.

prakhar1144 commented:
* I implemented all of the above proposed UX except:
  > Let's try having the "Watch later" button mark the onboarding step as done (so it won't pop up the video again). Instead, it can trigger a Welcome Bot message in 24 hrs (exact timing TBD):
  
  Posted a technical question in [#backend > onboarding video @ 💬](https://chat.zulip.org/#narrow/channel/3-backend/topic/onboarding.20video/near/2114780)

* https://github.com/zulip/zulip/pull/32074#issuecomment-2695279422
  > I dunno if it's critical, but why does clicking on a future time in the time bar not work for jumping to that time?
  This works for me. In which browser you're trying ?


* Welcome bot message:
  <img width="897" alt="Screenshot 2025-03-06 at 11 04 56 PM" src="https://github.com/user-attachments/assets/49341a9b-a7e0-4186-a0d8-be9d66e7c130" />
  
* Video ended buttons:
<img width="1026" alt="Screenshot 2025-03-06 at 11 32 53 PM" src="https://github.com/user-attachments/assets/0cc56e0a-0136-4cd0-ad81-0a852443506e" />

alexmv commented on zproject/default_settings.py:
```diff
@@ -708,3 +708,7 @@
+
+# URL of the onboarding video displayed to new users.
+# Set it to None to disable it.
+ONBOARDING_VIDEO_URL: str | None = "https://static.zulip.com/static/onboarding-video/zulip-10.mp4"
```
`static.zulipchat.com`, not `static.zulip.com`

prakhar1144 replied:
Yeah, will update with https://github.com/zulip/zulip/pull/32074#discussion_r1977942708

This is not currently being used.

prakhar1144 replied:
Done

alya commented:
> This works for me. In which browser you're trying ?

Chrome on a Mac. 

![2025-03-06 21 57 09](https://github.com/user-attachments/assets/729eded9-ac9e-48c3-bcf2-4c77102bedb2)

alya commented:
It it possible to make the "Let's go!" button show up even if you've full-screened the video? It currently does not.

alya commented:
Whoa, why is part of the Welcome Bot message missing when I make a  new user in the new dev realm?

![Screenshot 2025-03-06 at 22 20 23@2x](https://github.com/user-attachments/assets/7718a629-da78-40cb-8abb-839783a01784)

It's fine when I create a new realm (so the user is an owner):

![Screenshot 2025-03-06 at 22 19 59@2x](https://github.com/user-attachments/assets/d4d1f91a-b9d8-43de-816c-6fc22c3597b1)

alya commented:
Just as a reminder, this is not the final video, so we should update the link at some point.

prakhar1144 commented:
> Chrome on a Mac.

Tried out on Chrome, Brave & Safari


https://github.com/user-attachments/assets/959e621d-cdf5-4259-afe3-8c96d32d8df4

amanagr commented:
![Screenshot 2025-03-11 113455](https://github.com/user-attachments/assets/867865c5-eda1-463b-a106-c2a311762c34)
![image](https://github.com/user-attachments/assets/85357791-98a0-4074-a245-4c0275a1bd7a)


Made the background dim and changed the button color.

alya commented:
We discussed it [briefly here ](https://chat.zulip.org/#narrow/channel/101-design/topic/Mark.20onboarding.20messages.20as.20unread/near/1832305)[#design > Mark onboarding messages as unread @ 💬](https://chat.zulip.org/#narrow/channel/101-design/topic/Mark.20onboarding.20messages.20as.20unread/near/1832305) & I remember in one of the GH comment it was discussed that the intent was to let the default realm populated in dev to act as old orgs and we can test these new onboarding messages etc via creating new realm.

OK, I dunno, it's super confusing if creating a new user in dev doesn't look like creating a new user in production. Yes, it's a new user in an existing org, so it should replicate that experience...

prakhar1144 commented:
> OK, I dunno, it's super confusing if creating a new user in dev doesn't look like creating a new user in production. Yes, it's a new user in an existing org, so it should replicate that experience...

Okay, we can improve the dev experience. I'll extend that thread to discuss.

prakhar1144 commented:
Updates in the recent push:
* Watch later button schedules a reminder.
* Use `https://static.zulipchat.com/static/onboarding-video/zulip-10.mp4` instead of including it with the release tarball approach.
* Exit fullscreen when video ends -- helps to make "Let's go!" button visible.

timabbott commented on web/src/onboarding_steps.ts:
```diff
@@ -57,10 +75,95 @@ function narrow_to_dm_with_welcome_bot_new_user(
+                    e.preventDefault();
+                    e.stopPropagation();
+                    // Schedule a reminder message.
+                    const reminder_delay_seconds = 30;
```
This needs an actual value. Probably 48 hours?

prakhar1144 replied:
Yeah, [30 was for testing](https://github.com/zulip/zulip/pull/32074#issuecomment-2716641651).

Changed it to:
> const reminder_delay_seconds = 48 * 60 * 60;

timabbott commented on web/styles/modal.css:
```diff
@@ -667,3 +667,67 @@
+        }
+    }
+
+    video {
```
This should have a class on it; we don't want to ever write CSS for bare elements. (It has performance and readability/greppability issues).

timabbott commented on zerver/actions/scheduled_messages.py:
```diff
@@ -104,14 +108,15 @@ def do_schedule_messages(
+                    scheduled_message.to_dict() for scheduled_message, ignored in scheduled_messages
+                ],
+            }
+            send_event_on_commit(sender.realm, event, [sender.id])
```
What's the reason for skipping events?

prakhar1144 replied:
Sender is welcome bot here. No need to send the event to welcome bot.

timabbott commented on zerver/lib/events.py:
```diff
@@ -572,6 +572,9 @@ def fetch_initial_state_data(
         )
 
+        if settings.ONBOARDING_VIDEO_URL is not None:
+            state["onboarding_video_url"] = settings.ONBOARDING_VIDEO_URL
```
I think we want `None` values to be present here; we don't do optional keys in `register` much.

prakhar1144 replied:
Removed the `if settings.ONBOARDING_VIDEO_URL is not None:` conditional.

Is there any benefit to include such fields which don't have any value ? The response data will be slightly smaller in size (not much) if we just drop such fields.

timabbott replied:
It can be simpler for clients to parse a response with a more predictable set of fields in the response.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -14843,6 +14843,14 @@ paths:
+                        description: |
+                          URL of the onboarding video displayed to new users.
+
+                          Present unless disabled by the server administrators in `/etc/zulip/settings.py`.
```
I think we probably want "Present if `onboarding_steps` is present in `fetch_event_types`." to be the behavior; i.e., key this under a `want("onboarding_steps")` check in `zerver/lib/events.py. 

(The alternative is `realm`, but given the coupling with onboarding, this seems right).

prakhar1144 replied:
Made the changes to make it present if `onboarding_steps` is present in `fetch_event_types`

timabbott commented on zerver/views/onboarding_steps.py:
```diff
@@ -1,23 +1,60 @@
+
+        realm = user.realm
+        sender = get_system_bot(settings.WELCOME_BOT, realm.id)
+        client = RequestNotes.get_notes(request).client
```
Do we have other system bots messages use this as a client? `_internal_prep_message` uses "Internal" as the client, which seems more correct.

prakhar1144 replied:
Umm, No. But I thought this goes in the category of "message scheduled by the user on behalf of welcome bot" -- so client should be noted accordingly.

But we can use "Internal" too -- some user action resulted in internal bot message which was not instant but scheduled.

Changed it to Internal.

timabbott commented on zproject/prod_settings_template.py:
```diff
@@ -886,3 +886,7 @@
+
+## URL of the onboarding video displayed to new users.
+## Set it to None to disable it.
+# ONBOARDING_VIDEO_URL: str | None = "https://static.zulipchat.com/static/onboarding-video/zulip-10.mp4"
```
We don't generally include type annotations in comments in this field. Did a bit of a tweak to do the documentation too.
``` diff
diff --git a/zproject/prod_settings_template.py b/zproject/prod_settings_template.py
index be9d8506a2..655bb56cd2 100644
--- a/zproject/prod_settings_template.py
+++ b/zproject/prod_settings_template.py
@@ -887,6 +887,9 @@ CAMO_URI = "/external_content/"
 ## Directory containing Markdown files for the server's policies.
 # POLICIES_DIRECTORY = "/etc/zulip/policies/"
 
-## URL of the onboarding video displayed to new users.
-## Set it to None to disable it.
-# ONBOARDING_VIDEO_URL: str | None = "https://static.zulipchat.com/static/onboarding-video/zulip-10.mp4"
+## URL of the onboarding video to show to new users. You can use this
+## to host the official video on your network, or to provide your own
+## introductory video with details on how your organization uses Zulip.
+##
+## A value of None disables the onboarding video experience.
+# ONBOARDING_VIDEO_URL = "https://static.zulipchat.com/static/onboarding-video/zulip-10.mp4"
```

prakhar1144 replied:
Renamed `ONBOARDING_VIDEO_URL` to `NAVIGATION_TOUR_VIDEO_URL` (and related minor comment update).

prakhar1144 commented on web/styles/modal.css:
```diff
@@ -667,3 +667,67 @@
+        width: 100%;
+        visibility: hidden;
+
+        > button {
```
Used `#navigation-tour-video-ended-button`

prakhar1144 commented:
Updates in the recent push:
* Updated the whole PR to align with our decision to call it `navigation_tour_video` instead of `onboarding_video`.
* Addressed review comments.
* Three tweaks decided in CZO: [#design > onboarding video: modal @ 💬](https://chat.zulip.org/#narrow/channel/101-design/topic/onboarding.20video.3A.20modal/near/2121976)
* https://github.com/zulip/zulip/pull/32074#issuecomment-2718525061

<img width="592" alt="Screenshot 2025-03-13 at 7 49 23 PM" src="https://github.com/user-attachments/assets/7420fe96-93d2-462f-8626-6bc25efb556e" />

<img width="524" alt="Screenshot 2025-03-13 at 7 00 29 PM" src="https://github.com/user-attachments/assets/1070dfb0-daee-4edf-86cd-379abcf5c2cb" />

timabbott commented on web/templates/navigation_tour_video_modal.hbs:
```diff
@@ -4,6 +4,6 @@
     </video>
     <div id="navigation-tour-video-ended-button-wrapper">
-        <button id="navigation-tour-video-ended-button">{{t "Let's go!"}}</button>
+        <button id="navigation-tour-video-ended-button" class="action-button-primary-brand">{{t "Let's go!"}}</button>
```
I feel like this should be using this:

```
+        {{> components/action_button
+          label=(t "Let's go!")
+          id="navigation-tour-video-ended-button"
+          attention="primary"
+          intent="brand"
+        }}
```

But visually it looks good, and I've not audited what CSS we should delete doing that. Perhaps you can @sayamsamal can collaborate on cleaning that up.
